src/shmem_team.c

```diff
@@ -0,0 +1,455 @@
+                                      int PE_size, const shmemx_team_config_t *config, long config_mask,
+                                      shmem_internal_team_t **new_team)
+{
+
+    *new_team = SHMEMX_TEAM_INVALID;
+
+    if (parent_team == SHMEMX_TEAM_INVALID) {
+        return 0;
+    }
+
+    int global_PE_start = shmem_internal_team_pe(parent_team, PE_start);
+    int global_PE_end   = global_PE_start + PE_stride * (PE_size -1);
+
+    if (PE_size <= 0 || PE_stride < 1 || global_PE_start >= shmem_internal_num_pes ||
+        global_PE_end >= shmem_internal_num_pes) {
+        RAISE_WARN_STR("Invalid start, stride, or size in team_split operation");
+        return -1;
+    }
+
+    shmem_internal_team_t *myteam = calloc(1, sizeof(shmem_internal_team_t));
+
+    myteam->start       = global_PE_start;
+    myteam->stride      = PE_stride;
+    myteam->size        = PE_size;
+    if (config) {
+        myteam->config      = *config;
+        myteam->config_mask = config_mask;
+    }
+    myteam->contexts_len = 0;
+    myteam->psync_idx = 0;
+
+    myteam->my_pe = shmem_internal_pe_in_active_set(shmem_internal_my_pe, myteam->start, PE_stride, PE_size);
+
+    if (myteam->my_pe != -1) {
+
+        size_t psync = shmem_internal_team_choose_psync(parent_team, REDUCE);
+
+        shmem_internal_op_to_all(psync_pool_avail_reduced,
+                                 psync_pool_avail, N_PSYNC_BYTES, 1,
+                                 myteam->start, PE_stride, PE_size, NULL,
+                                 &shmem_internal_psync_pool[psync],
+                                 SHM_INTERNAL_BAND, SHM_INTERNAL_UCHAR);
+
+        shmem_internal_team_release_psyncs(parent_team, REDUCE);
+
+        /* Select the least signficant nonzero bit, which corresponds to an available pSync. */
+        myteam->psync_idx = shmem_internal_bit_1st_nonzero(psync_pool_avail_reduced, N_PSYNC_BYTES);
+        if (myteam->psync_idx == -1 || myteam->psync_idx >= shmem_internal_params.TEAMS_MAX) {
+            RAISE_WARN_MSG("No more teams available (max = %ld), try increasing SHMEM_TEAMS_MAX\n",
+                            shmem_internal_params.TEAMS_MAX);
+            myteam->psync_idx = -1;
+        } else {
+            /* Set the selected psync bit to 0, reserving that slot */
+            shmem_internal_bit_clear(psync_pool_avail, N_PSYNC_BYTES, myteam->psync_idx);
+
+            for (size_t i = 0; i < N_PSYNCS_PER_TEAM; i++)
+                myteam->psync_avail[i] = 1;
+
+            *new_team = myteam;
+
+            shmem_internal_team_pool[myteam->psync_idx] = *new_team;
+        }
+    }
+
+    size_t psync = shmem_internal_team_choose_psync(parent_team, SYNC);
+
+    shmem_internal_barrier(parent_team->start, parent_team->stride, parent_team->size,
+                           &shmem_internal_psync_barrier_pool[psync]);
+
+    shmem_internal_team_release_psyncs(parent_team, SYNC);
+
+    if (myteam->psync_idx == -1)
```

Comment:
```suggestion
    if (my_pe >= 0 && myteam->psync_idx == -1)
```

Reply:
Suggest a more specific error check here.

Reply:
We won't be able to check `myteam->psync_dx` here after fixing the leak of `myteam`.  Any suggestions there?  For now I'm setting myteam to NULL outside the loop and checking that here as well...

Reply:
@jdinan - please review that change in https://github.com/Sandia-OpenSHMEM/SOS/pull/886/commits/daff427554a791de1a4c59a5848a091b77ad14e7